### DISCUSSION TITLE
如何重新触发render方法

### DISCUSSION THREAD

**Discussion by yl196356303** [umijs/umi#9717] [Q&A]
![image](https://user-images.githubusercontent.com/55053022/200304677-e709f06f-7b56-434d-9863-00a2b2c30ed9.png)

**Comment by yl196356303** [comment]
我想在登录成功后请求服务器动态渲染路由，请问需要怎么做

**Reply by a992681784** [reply]
请问下 现在找到解决方案了吗？我也遇到同样的需求了。app.tsx只会刚进页面执行一次，点击登录后不会再执行了。

**Reply by yl196356303** [reply]
登录时手动调用patchClientRoutes

**Reply by a992681784** [reply]
好的 谢谢 我试下 是从app.tsx里引用这个patchClientRoutes方法进行调用吗？

**Reply by echo-machile** [reply]
> 我想在登录成功后请求服务器动态渲染路由，请问需要怎么做

按照下面的方法，我的antd布局被覆盖掉了，遇到过这种情况嘛？

**Reply by jaxsh** [reply]
> 登录时手动调用patchClientRoutes

能详细说一下登录的时候如何手动调用吗？谢谢

**Comment by yl196356303** [comment]
手动调用oldRender()方法直接报错

**Comment by fz6m** [comment]
```tsx
// src/app.tsx

import React from 'react'

const NotFound = () => <div>404</div>
const Wrapper = ({ children }: any) => (
  <React.Suspense>{children}</React.Suspense>
)
const Add = React.lazy(() => import('./components/Add'))

const MAPS: any = {
  '/add': (
    <Wrapper>
      <Add />
    </Wrapper>
  ),
}

let roleRoutes: any[] = []
const getRoleRoutes = () => {
  return new Promise<void>((resolve) => {
    setTimeout(() => {
      roleRoutes = ['/add']
      resolve()
    }, 1000)
  })
}

export const patchClientRoutes = ({ routes }: any) => {
  routes[0].routes.unshift(
    ...roleRoutes.map((path: string) => ({
      id: path,
      path,
      element: MAPS?.[path] || <NotFound />,
    }))
  )
}

export const render = async (oldRender: Function) => {
  // 如果请求太慢，可选：自己实现一个加载器效果
  document.querySelector('#root')!.innerHTML = `<div>loading...</div>` 

  await getRoleRoutes()
  oldRender()
}
```

**Reply by yl196356303** [reply]
大佬我试一下还是不行呀，并不能在我登录成功的时候去服务器请求权限，触发render方法，重新生成路由

**Reply by fz6m** [reply]
如果已经 render 出登录页了，就不可以修改路由了（除非刷新页面，此时在 render 前带着有权的 cookie 请求路由表），需要在 render 前做好准备。

如果需要在 render 后进行限制，可以考虑在 Layout 校验权限。

**Reply by yl196356303** [reply]
好吧，那前端只能写死一份路由表了

**Reply by fz6m** [reply]
正常来说，有 sso 分配 cookie 权限，在访问某个系统时，render 前，请求拿到这个 cookie 权限下面所有的路由表，是这么个流程。

**Reply by yl196356303** [reply]
大佬patchClientRoutes修改路由无效果。。。
![image](https://user-images.githubusercontent.com/55053022/201882417-91940034-6866-4b3c-8c11-a0d7ce1ba1fd.png)
我过滤掉某个路由，依旧可以访问这个路由

**Reply by fz6m** [reply]
给一个最小复现看看。

**Reply by yl196356303** [reply]
似乎只能修改routes，比如push，pop，shift等方法，不能直接覆盖routes
![image](https://user-images.githubusercontent.com/55053022/201885400-d1c493b7-b731-415b-9d3f-5f5015144173.png)

**Reply by yl196356303** [reply]
不清楚是否算bug。。。

**Reply by yl196356303** [reply]
https://codesandbox.io/s/eager-archimedes-8m3hmc?file=/src/app.tsx

**Reply by fz6m** [reply]
这个地方的 `routes` 对象只是传进来，只能改对象内的引用，全改了无效。

```ts
  for (let i = 0; i < routes.length; i++) {
    if (routes[i].path === '/about') {
      delete routes[i];
      break;
    }
  }
```

**Reply by yl196356303** [reply]
大佬，请教一下这种写法为啥报错
![image](https://user-images.githubusercontent.com/55053022/201909612-26801d43-1c23-4235-a01d-d1f6e8cdb861.png)
 Functions are not valid as a React child. This may happen if you return a Component instead of <Component /> from render. Or maybe you meant to call this function rather than return it.

**Reply by yl196356303** [reply]
在线demo可以进去看到

**Reply by fz6m** [reply]
看一下我最上面给的代码例子，你这个写法 children 不是组件。

**Reply by yl196356303** [reply]
这个就是参考你的demo来的，然后就报了上面那个错。。。，使用createElement(require(path).default)没问题

**Reply by fz6m** [reply]
仔细看看。有区别。

**Reply by waittingbar** [reply]
> 大佬，请教一下这种写法为啥报错 ![image](https://user-images.githubusercontent.com/55053022/201909612-26801d43-1c23-4235-a01d-d1f6e8cdb861.png) Functions are not valid as a React child. This may happen if you return a Component instead of from render. Or maybe you meant to call this function rather than return it.

我发现写了React.createElement之后，打包体积变大很多。。

**Reply by echo-machile** [reply]
> ```tsx
> // src/app.tsx
> 
> import React from 'react'
> 
> const NotFound = () => <div>404</div>
> const Wrapper = ({ children }: any) => (
>   <React.Suspense>{children}</React.Suspense>
> )
> const Add = React.lazy(() => import('./components/Add'))
> 
> const MAPS: any = {
>   '/add': (
>     <Wrapper>
>       <Add />
>     </Wrapper>
>   ),
> }
> 
> let roleRoutes: any[] = []
> const getRoleRoutes = () => {
>   return new Promise<void>((resolve) => {
>     setTimeout(() => {
>       roleRoutes = ['/add']
>       resolve()
>     }, 1000)
>   })
> }
> 
> export const patchClientRoutes = ({ routes }: any) => {
>   routes[0].routes.unshift(
>     ...roleRoutes.map((path: string) => ({
>       id: path,
>       path,
>       element: MAPS?.[path] || <NotFound />,
>     }))
>   )
> }
> 
> export const render = async (oldRender: Function) => {
>   // 如果请求太慢，可选：自己实现一个加载器效果
>   document.querySelector('#root')!.innerHTML = `<div>loading...</div>` 
> 
>   await getRoleRoutes()
>   oldRender()
> }
> ```

这种方式的动态路由会覆盖掉全局布局吧？

**Reply by fz6m** [reply]
给最小复现看看

**Comment by yl196356303** [comment]
二次调用render方法会直接报错

**Comment by yl196356303** [comment]
不过我尝试了一下似乎可以手动调用patchClientRoutes方法

**Reply by waittingbar** [reply]
你是这么在外面调用patchClientRoutes方法的啊

**Reply by yl196356303** [reply]
export 导出 import 引入

**Comment by jaxsh** [comment]
打开登陆页面的时候会调用render，patchClientRoutes，此时因为还未登录，并未拉取路由数据
登录然后动态拉取路由数据，但是页面没有触发render，patchClientRoutes。导致点击菜单404。
此时重新刷新页面可以可以触发render，patchClientRoutes，之后点击菜单就不会404了。但是此方式不太优雅，不知道有没有更好的方式

**Reply by fz6m** [reply]
一般情况是接口拦截器发现没权，直接带着 redirect url 跳去 sso ，登录完再被 sso 重定向回来，这样本质上是又访问回来一次了，体验是正常的。

**Reply by cxdmaye** [reply]
大佬，有没有办法覆盖redirect呢？

**Reply by fz6m** [reply]
不明白你要覆盖哪个，需要给一个最小复现

**Reply by cxdmaye** [reply]
就是覆盖路由表里最外层那个保底的
```
{
    path: '/*',
    redirect: '/home',
}
```
希望根据后端拉取到的动态路由找到第一个有权限的页面，把对应的path覆盖上面的 redirect 中 '/home'。
// 防止 /home也没权限，假设。

尝试了两种方式，
push，在` routes[0].routes` 中 push一个新的路由，需要处理element，没搞懂怎么写这部分
edit, 找到 上面这一条，修改他的redirect字段，但只找到了一个 `element.props.to` 看起来像，修改的话报只读属性无法修改。


最小复现明天提供

**Reply by Icodezhan** [reply]
> 一般情况是接口拦截器发现没权，直接带着 redirect url 跳去 sso ，登录完再被 sso 重定向回来，这样本质上是又访问回来一次了，体验是正常的。

可是没有sso的情况呢？

**Reply by fz6m** [reply]
如有问题，需要提供一个最小复现。

**Comment by tjj727** [comment]
打包后发现体积大了很多，之前大概6M，修改后8M

**Reply by fz6m** [reply]
自己用 [`ANALYZE`](https://umijs.org/docs/guides/env-variables#analyze) 分析下。
